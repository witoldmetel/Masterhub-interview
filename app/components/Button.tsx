import React from 'react'
import { Text, StyleSheet, Pressable } from 'react-native'

type ButtonProps = {
	title: string
	disabled?: boolean
	testID?: string

	onPress: () => void
}

export function Button({ onPress, title, disabled, testID }: ButtonProps) {
	return (
		<Pressable
			style={({ pressed }) => [
				styles.button,
				{ backgroundColor: disabled ? '#333' : '#ff4e2b', opacity: pressed ? 0.65 : 1 },
			]}
			onPress={onPress}
			disabled={disabled}
			testID={testID}>
			<Text style={styles.text}>{title}</Text>
		</Pressable>
	)
}

const styles = StyleSheet.create({
	button: {
		alignItems: 'center',
		borderRadius: 8,
		elevation: 3,
		justifyContent: 'center',
		marginVertical: 12,
		paddingHorizontal: 32,
		paddingVertical: 12,
	},
	text: {
		color: '#010103',
		fontSize: 16,
		fontWeight: 'bold',
		letterSpacing: 0.25,
		lineHeight: 21,
	},
})
