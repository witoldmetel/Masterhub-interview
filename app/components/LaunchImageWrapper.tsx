import React from 'react'
import { Dimensions, View, StyleSheet, Image } from 'react-native'
import Carousel from 'react-native-reanimated-carousel'

type LaunchImageWrapperProps = {
	links: {
		flickr_images?: string[]
		video_link: string
	}
}

export function LaunchImageWrapper({ links }: LaunchImageWrapperProps) {
	const width = Dimensions.get('window').width

	return (
		<View style={styles.imageContainer} testID="image-container">
			{links.flickr_images && links.flickr_images.length > 0 ? (
				<Carousel
					loop
					width={width}
					height={width / 2}
					autoPlay={true}
					data={links.flickr_images}
					scrollAnimationDuration={2000}
					renderItem={({ item }: { item: string }) => {
						return (
							<View style={styles.imageItem} testID="image">
								<Image key={item} source={{ uri: item }} style={styles.image} />
							</View>
						)
					}}
					testID="carousel"
				/>
			) : (
				<Image
					source={{ uri: `https://img.youtube.com/vi/${links.video_link.split('v=')[1]}/0.jpg` }}
					style={styles.videoThumbnail}
					testID="video-thumbnail"
				/>
			)}
		</View>
	)
}

const styles = StyleSheet.create({
	image: {
		height: 200,
		marginBottom: 8,
		resizeMode: 'contain',
		width: '100%',
	},
	imageContainer: {
		marginVertical: 16,
	},
	imageItem: {
		alignItems: 'center',
		justifyContent: 'center',
	},
	videoThumbnail: {
		borderRadius: 8,
		height: 200,
		resizeMode: 'cover',
		width: '100%',
	},
})
