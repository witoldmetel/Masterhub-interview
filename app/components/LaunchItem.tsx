import React from 'react'
import { Text, View, StyleSheet, Pressable } from 'react-native'
import { useNavigation } from '@react-navigation/native'
import { format } from 'date-fns'

import { LaunchType } from '../types'
import { LaunchDetailsScreenNavigationProp } from '../navigators/types'

type LaunchItemProps = {
	item: LaunchType
}

export const LaunchItem = ({ item }: LaunchItemProps) => {
	const navigation = useNavigation<LaunchDetailsScreenNavigationProp>()

	const onItemPress = (itemId: string) => {
		navigation.navigate('LaunchDetails', {
			launchId: itemId,
		})
	}

	return (
		<Pressable
			style={({ pressed }) => [styles.launchItem, { opacity: pressed ? 0.75 : 1 }]}
			onPress={() => onItemPress(item.id)}
			testID={item.id}>
			<Text style={styles.launchDate}>{format(new Date(item.launch_date_local), 'MMMM dd, yyyy')}</Text>
			<Text style={styles.launchMissionName}>{item.mission_name}</Text>
			<View
				style={[
					styles.launchRocketNameWrapper,
					{ backgroundColor: item.rocket.rocket.active ? '#38373c' : '#ff4e2b' },
				]}>
				<Text style={[styles.launchRocketName, { color: item.rocket.rocket.active ? '#fff' : '#1c1c1e' }]}>
					{item.rocket.rocket.name}
				</Text>
			</View>
		</Pressable>
	)
}

const styles = StyleSheet.create({
	launchDate: { color: '#ff4e2b', fontSize: 16, fontWeight: '500' },
	launchItem: {
		backgroundColor: '#1c1c1e',
		borderRadius: 16,
		borderWidth: 1,
		padding: 16,
	},
	launchMissionName: { color: '#fff', fontSize: 20, fontWeight: '500', marginVertical: 8 },
	launchRocketName: { color: '#fff', fontSize: 14, fontWeight: '500' },
	launchRocketNameWrapper: {
		alignSelf: 'flex-start',
		backgroundColor: '#38373c',
		borderRadius: 8,
		padding: 6,
	},
})
