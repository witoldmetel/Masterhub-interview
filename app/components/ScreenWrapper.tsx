import React from 'react'
import { StyleSheet } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

type ScreenWrapperProps = {
	children?: React.ReactNode
}

export const ScreenWrapper = ({ children }: ScreenWrapperProps) => {
	return (
		<SafeAreaView style={styles.rootContainer} edges={['top']} testID="root-container">
			{children}
		</SafeAreaView>
	)
}

const styles = StyleSheet.create({
	rootContainer: {
		backgroundColor: '#010103',
		flex: 1,
	},
})
