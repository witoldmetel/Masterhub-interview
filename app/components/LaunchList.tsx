import React from 'react'
import { Text, View, StyleSheet, FlatList } from 'react-native'

import { useLaunches } from '../hooks/useLaunches'
import { LaunchItem } from './LaunchItem'
import { Button } from './Button'
import { Loader } from './Loader'

type LaunchListProps = {
	searchQuery: string
}

export const LaunchList = ({ searchQuery }: LaunchListProps) => {
	const { launches, loadingLaunches, moreResults, handleLoadMore } = useLaunches(searchQuery)

	if (loadingLaunches) {
		return <Loader />
	}

	return (
		<>
			<FlatList
				data={launches}
				renderItem={({ item }) => <LaunchItem item={item} />}
				keyExtractor={item => item.id}
				ItemSeparatorComponent={() => <View style={styles.separator} />}
				ListEmptyComponent={
					<View style={styles.emptyContainer} testID="empty">
						<Text style={styles.emptyContainerText}>No results</Text>
					</View>
				}
				testID="flatlist"
			/>
			{moreResults ? (
				<Button onPress={handleLoadMore} title="Load More" disabled={loadingLaunches} testID="loadmore" />
			) : null}
		</>
	)
}

const styles = StyleSheet.create({
	emptyContainer: {
		alignItems: 'center',
		backgroundColor: '#f9c7a4',
		borderRadius: 16,
		borderWidth: 1,
		justifyContent: 'center',
		padding: 16,
	},
	emptyContainerText: {
		color: '#a1673f',
		fontSize: 24,
		fontWeight: '500',
	},
	separator: {
		height: 12,
	},
})
