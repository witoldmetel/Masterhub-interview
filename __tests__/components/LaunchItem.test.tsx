import React from 'react'
import renderer, { act } from 'react-test-renderer'

import { LaunchItem } from '../../app/components/LaunchItem'

const mockNavigation = { navigate: jest.fn() }
jest.mock('@react-navigation/native', () => ({
	useNavigation: () => mockNavigation,
}))

describe('LaunchItem', () => {
	let launch, launchItem, launchItemInstance, pressable

	beforeEach(() => {
		launch = {
			id: 'test-id',
			mission_name: 'Test Mission Name',
			launch_date_local: new Date('2023-03-28T22:00:00Z'),
			rocket: {
				rocket: {
					id: 'test-rocket-id',
					name: 'Test Rocket Name',
					active: true,
				},
			},
		}
		launchItem = renderer.create(<LaunchItem item={launch} />)
		launchItemInstance = launchItem.root
		pressable = launchItemInstance.findByProps({ testID: 'test-id' })
	})

	it('renders correctly', () => {
		const tree = launchItem.toJSON()

		expect(tree).toMatchSnapshot()
	})

	it('calls navigate with the correct arguments when pressed', () => {
		act(() => {
			pressable.props.onPress()
		})

		expect(mockNavigation.navigate).toHaveBeenCalledWith('LaunchDetails', {
			launchId: launch.id,
		})
	})
})
