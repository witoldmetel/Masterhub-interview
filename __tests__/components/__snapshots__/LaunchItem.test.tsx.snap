// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchItem renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#1c1c1e",
        "borderRadius": 16,
        "borderWidth": 1,
        "padding": 16,
      },
      Object {
        "opacity": 1,
      },
    ]
  }
  testID="test-id"
>
  <Text
    style={
      Object {
        "color": "#ff4e2b",
        "fontSize": 16,
        "fontWeight": "500",
      }
    }
  >
    March 29, 2023
  </Text>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontSize": 20,
        "fontWeight": "500",
        "marginVertical": 8,
      }
    }
  >
    Test Mission Name
  </Text>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "backgroundColor": "#38373c",
          "borderRadius": 8,
          "padding": 6,
        },
        Object {
          "backgroundColor": "#38373c",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontSize": 14,
            "fontWeight": "500",
          },
          Object {
            "color": "#fff",
          },
        ]
      }
    >
      Test Rocket Name
    </Text>
  </View>
</View>
`;
